### PR TITLE
Organize bridge rooms into a space

### PR DESCRIPTION
**Pull request by HelderFSFerreira**
Features:
- Create a space for each user
- Add all the rooms to the bridge
- Config flag to enable it

To do:
- Add avatar do space
- Manage space permissions

### CODE REVIEW

**Comment by sumnerevans** [comment]
You can add `Co-authored-by: Other User <other.user@email.com>` to the commit to credit other contributors.

File: main.go [hunk 1]
```diff
@@ -179,8 +179,9 @@ type Bridge struct {
 	usersByMXID         map[id.UserID]*User
 	usersByUsername     map[string]*User
 	usersLock           sync.Mutex
-	managementRooms     map[id.RoomID]*User
+	spaceRooms          map[id.RoomID]*User
 	managementRoomsLock sync.Mutex
+	managementRooms     map[id.RoomID]*User
```
Comment (sumnerevans) [inline main.go:184]: Move this back above `managementRoomsLock`

File: portal.go [hunk 2]
```diff
@@ -1176,6 +1178,15 @@ func (portal *Portal) CreateMatrixRoom(user *User, groupInfo *types.GroupInfo, i
 	return nil
 }
 
+func (portal *Portal) addToSpace(spaceID id.RoomID, portalID id.RoomID) {
+
```
Comment (sumnerevans) [inline portal.go:1182]:
```suggestion
```

File: portal.go [hunk 3]
```diff
@@ -1176,6 +1178,15 @@ func (portal *Portal) CreateMatrixRoom(user *User, groupInfo *types.GroupInfo, i
...
+	parentSpaceContent := make(map[string]interface{})
+	parentSpaceContent["via"] = []string{"matrix.wounn.xyz"}
+
+	portal.log.Errorln("adding room " + portalID + " to the space " + spaceID)
```
Comment (sumnerevans) [inline portal.go:1186]:
```suggestion
	portal.log.Debugfln("adding room %s to the space %s", portalID, spaceID)
```

**Comment by clmnin** [comment]
@sumnerevans this feature is not complete. 
We need to 
1. Help current users migrate / migration script
2. Test for federation (currently `via` is hard-coded)
3. Discuss the best way to set this in config.yaml
etc.

Could you review the code and share your feedback?

File: portal.go [hunk 4]
```diff
@@ -1176,6 +1178,15 @@ func (portal *Portal) CreateMatrixRoom(user *User, groupInfo *types.GroupInfo, i
 	return nil
 }
 
+func (portal *Portal) addToSpace(spaceID id.RoomID, portalID id.RoomID) {
```
Comment (sumnerevans) [inline portal.go:1181]: This should also set the space's parent. (`m.space.parent`, or whatever they are calling it)

File: portal.go [hunk 5]
```diff
@@ -1176,6 +1178,15 @@ func (portal *Portal) CreateMatrixRoom(user *User, groupInfo *types.GroupInfo, i
...
+
+	parentSpaceContent := make(map[string]interface{})
+	parentSpaceContent["via"] = []string{"matrix.wounn.xyz"}
```
Comment (sumnerevans) [inline portal.go:1184]: This needs to be the bridge's homeserver URL

File: portal.go [hunk 6]
```diff
@@ -1141,6 +1141,8 @@ func (portal *Portal) CreateMatrixRoom(user *User, groupInfo *types.GroupInfo, i
 	portal.ensureUserInvited(user)
 	user.syncChatDoublePuppetDetails(portal, true)
 
+	portal.addToSpace(user.getSpaceRoom(), portal.MXID)
```
Comment (sumnerevans) [inline portal.go:1144]: Only do this if enabled in the config

File: user.go [hunk 7]
```diff
@@ -179,6 +179,53 @@ func (bridge *Bridge) NewUser(dbUser *database.User) *User {
...
+	}
+	return roomID
+}
+
+func (user *User) setSpaceRoom(spaceID id.RoomID) {
+	existingUser, ok := user.bridge.spaceRooms[spaceID]
+	if ok {
+		existingUser.SpaceRoom = ""
+		existingUser.Update()
+	}
```
Comment (sumnerevans) [inline user.go:222]: What is this actually accomplishing? Aren't you overriding it immediately afterwards?

**Review by sumnerevans — CHANGES_REQUESTED** [review]
Thanks for working on this! Generally seems pretty good. Looks like there's some TODOs still around config that also need to be addressed.

File: portal.go (hunk 4 above)
Comment (clmnin) [reply portal.go:1181]: We can choose one to two ways - either `m.space.child` or `m.space.parent`
From [MSC 1772](https://github.com/matrix-org/matrix-doc/blob/old_master/proposals/1772-groups-as-rooms.md#membership-of-spaces)
> The parent/child relationship can be expressed in one of two ways

**Comment by sumnerevans** [comment]
As far as how to add to the config, I'd just do something like: `bridge.create_space_per_user` or something. Might have to think a bit about how to best make that work with multi-tenant bridges.

I think it may make sense to something like what the `ResendBridgeInfo` function does for when the user enables space creation when it was previously disabled/unspecified.

File: portal.go (hunk 5 above)
Comment (clmnin) [reply portal.go:1184]: Yup. Should have fixed that when this commit was pushed.

From MSC 1772
> `via` key which gives a list of candidate servers that can be used to join the room

It means these servers will be used to bootstrap the join request ([permalink](https://matrix.to/#/!LPqOcoJQuGkweiTedq:riot.ovh/$H5vDqjNncsKd-2K4OpD2nbteji_ZPnWmtjeaLd1FJRg?via=matrix.org&via=envs.net&via=mozilla.org) to comment in spaces:riot.ovh)

The bridge's homeserver should be part of via - yes. Should be include the user's homeserver too if it is different from the bridge's homeserver? @HelderFSFerreira we'll try testing this on a federated instance.

**Comment by clmnin** [comment]
`ResendBridgeInfo` is a good option to let a user migrate to spaces. Will explore that 👍🏻 

So migration will be
1. bridge.create_space_per_user
2. resend_bridge_info

New users can just set 
1. bridge.create_space_per_user

File: portal.go (hunk 4 above)
Comment (tulir) [reply portal.go:1181]: It's not an either or thing, both is the correct option for canonical relationships.

Comment (clmnin) [reply portal.go:1181]: @tulir So we need to run both 
1. `update space-room` by adding a child
2. `create a new room` by stating it's parent

I just checked element-web and only saw one request go through - `m.space.parent`. (Other requests are related to sync). Why does element-web only run te `m.space.parent` request?

File: main.go (hunk 1 above)
Comment (HelderFSFerreira) [reply main.go:184]: It was fixed on my last commit but it was needed to add spaceRoomsLock

File: portal.go (hunk 5 above)
Comment (HelderFSFerreira) [reply portal.go]: I've added the bridge home server as the value of the via.
Are we sure that we want to have the client's home server in there too?

File: portal.go (hunk 4 above)
Comment (HelderFSFerreira) [reply portal.go]: Done on 0fb2abb

**Comment by tulir** [comment]
Merged in 3a9b3ab41fcea2dd499bf285e719b316503c0f77

**Pull request closed without merging**